File: python/151-200/166. Fraction to Recurring Decimal.py
import collections
class Solution(object):
    def fractionToDecimal(self, numerator, denominator):
        """
        :type numerator: int
        :type denominator: int
        :rtype: str
        """
        res= ""
        if denominator == 0:
            return res
        if numerator == 0:
            return "0"
        if denominator < 0 and numerator < 0:
            denominator *= -1
            numerator *= -1
        if denominator < 0 or numerator < 0:
            numerator = abs(numerator)
            denominator = abs(denominator)
            res = "-" + res
        res_dict = collections.defaultdict(int)
        if numerator% denominator == 0:
            return res + str(numerator/denominator)
        else:
            res = res + str(numerator / denominator) + "."
            data = ""
            lenght = 0
            r = numerator % denominator
            while r:
                if res_dict[r]:
                    index = res_dict[r]
                    res = res[:index] + "(" + res[index:] + ")"
                    break
                res_dict[r] = len(res)
                r *= 10
                res += str( r/ denominator)
                r = r % denominator
        return res
            

File: cpp/151-200/166. Fraction to Recurring Decimal.cc
#include <iostream>
#include <string>
#include <vector>
#include <map>
#include <unordered_map>
using namespace std;
class Solution {
public:
    string fractionToDecimal(long long n, long long d) {
        if (n == 0)
            return "0";
        string res;
        if (n < 0 ^ d < 0)
            res += '-';
        n = abs(n);
        d = abs(d);
        res += to_string(n / d);
        if (n % d == 0)
            return res;
        res += ".";
        unordered_map<int, int> map;
        for (long long r = n % d; r; r %= d) {
            if (map.count(r) > 0) {
                res.insert(map[r], 1, '(');
                res += ")";
                break;
            }
            map[r] = res.size();
            r *= 10;
            res += to_string(r / d);
        }
        return res;
    }
};

int main()
{
    Solution sol = Solution();
    cout << sol.fractionToDecimal(1, 3) << endl;
}